### PR TITLE
Fix #227: Fix SSRC Mapping when people join a recorded conversation

### LINKED ISSUE #227: No entry in json files for people who join after recording has started
When someone joins a session which is recording, jetsi meet will not record his track in metadata.json nor endpoints.json.

### PR DESCRIPTION
**Pull request by roidelapluie**
Add the mapping from SSRC to EndpointID to the Synchronizer instance used by
RecorderRtpImpl when new SSRCs are signalled over colibri.

### CODE REVIEW

**Comment by roidelapluie** [comment]
It is signes now

**Comment by roidelapluie** [comment]
ping @gavllew

**Comment by gavllew** [comment]
I can't help much, as I'm just another third party - I can't approve pull requests.  However I do at least have a comment on the diff which I'll add in a second.

File: src/main/java/org/jitsi/videobridge/RtpChannel.java [hunk 1]
```diff
@@ -1857,13 +1857,38 @@ public void setSources(List<SourcePacketExtension> sources)
...
                     // Do allow the number of explicitly signalled SSRCs to
                     // exceed the limit.
                     addReceiveSSRC(addedSSRC, false);
+
+                    // If recording is enabled, we need to add the mapping from
+                    // SSRC to EndpointID into the Synchronizer
+                    // instance used by RecorderRtpImpl.
+                    if (addToSyncronizer)
+                    {
+                        synchronizer.setEndpoint(addedSSRC & 0xffffffffl,
```
Comment (gavllew) [inline src/main/java/org/jitsi/videobridge/RtpChannel.java]: The `synchronizer` variable will be out-of-scope before you reach this code block.

**Comment by bgrozev** [comment]
Looks OK to me, except that it doesn't compile :). Should be a quick fix.

Are you actually using some variant of this? I don't understand what problem it solves. If I understand correctly, it only makes a difference when recording SSRCs that are signaled, but don't send and RTP/RTCP packets.

**Comment by gavllew** [comment]
The core problem was that new streams added after recording had started were not appearing in metadata.json if the SSRCs were signalled before arriving in media packets.  This is because the metadata attempts to calculate the wall-clock instant that the stream starts, which depends on the per-Endpoint RTP timestamp mapping maintained by the Synchronizer.

Xura is using a variant of this fix, but we're currently unable to contribute anything because we've never managed to get the CLA sorted out. :-(

**Comment by bgrozev** [comment]
I see now that the code in acceptDataInputStreamDatagramPacket() only updates the synchronizer if the SSRC is new. Thanks for the explanation.

We have the CLA for @roidelapluie so we can get this merged when the PR is fixed.

**Comment by roidelapluie** [comment]
updated

File: src/main/java/org/jitsi/videobridge/RtpChannel.java (hunk 1 above)
Comment (roidelapluie) [inline src/main/java/org/jitsi/videobridge/RtpChannel.java]: fixed

File: src/main/java/org/jitsi/videobridge/RtpChannel.java [hunk 2]
```diff
@@ -1726,13 +1726,38 @@ public void setSources(List<SourcePacketExtension> sources)
         addedSSRCs.removeAll(oldSignaledSSRCs);
         if (!addedSSRCs.isEmpty())
         {
+
+            boolean addToSyncronizer = false;
+            Recorder recorder = getContent().getRecorder();
```
Comment (bgrozev) [inline src/main/java/org/jitsi/videobridge/RtpChannel.java]: Calling getRecorder will create a Recorder instance if one doesn't exist (regardless of whether recording is enabled or not). It should only be called if isRecording() returns true (yeah, the design of the API is not good...).

Comment (roidelapluie) [inline src/main/java/org/jitsi/videobridge/RtpChannel.java]: Fixed

**Comment by bgrozev** [comment]
@zbettenbuk this looks like it may solve one of the issues you described

**Comment by zbettenbuk** [comment]
@bgrozev which one you mean?

**Comment by bgrozev** [comment]
The missing entries from metadata.json

**Pull request merged**